gossip/components/requester.c

```diff
@@ -17,136 +16,184 @@
+  size = hash243_set_size(&state->transactions);
+  rw_lock_handle_unlock(&state->lock);
+
+  return size >= MAX_TX_REQ_NBR;
+}
+
+retcode_t requester_clear_request(requester_state_t *const state,
+                                  flex_trit_t const *const hash) {
+  if (state == NULL || hash == NULL) {
+    return RC_REQUESTER_COMPONENT_NULL_STATE;
+  }
+
+  rw_lock_handle_wrlock(&state->lock);
+  // TODO REMOVE
+  rw_lock_handle_unlock(&state->lock);
+
+  return RC_OK;
 }
 
 retcode_t request_transaction(requester_state_t *const state,
-                              trit_array_p const hash) {
+                              flex_trit_t *const hash,
+                              bool const is_milestone) {
   retcode_t ret = RC_OK;
   bool exists = false;
+  trit_array_t key = {.trits = hash,
+                      .num_trits = HASH_LENGTH_TRIT,
+                      .num_bytes = FLEX_TRIT_SIZE_243,
+                      .dynamic = 0};
 
-  if (state == NULL) {
+  if (state == NULL || hash == NULL) {
     return RC_REQUESTER_COMPONENT_NULL_STATE;
-  } else if (hash == NULL) {
-    return RC_REQUESTER_COMPONENT_NULL_HASH;
   }
 
-  // TODO(thibault) check null hash
+  //  TODO CHECK NULL HASH
   if ((ret = iota_tangle_transaction_exist(
-           state->tangle, TRANSACTION_FIELD_HASH, hash, &exists))) {
+           state->tangle, TRANSACTION_FIELD_HASH, &key, &exists)) != RC_OK) {
     return ret;
   }
   if (exists) {
     return RC_OK;
   }
-  if (CL_CONTAINS(state->list, hash) == false) {
-    if (requester_is_full(state) == false) {
-      if (CL_PUSH_BACK(state->list, hash) != CL_SUCCESS) {
-        log_warning(REQUESTER_COMPONENT_LOGGER_ID,
-                    "Adding new transaction request to the list failed\n");
-        return RC_REQUESTER_COMPONENT_FAILED_ADD_LIST;
-      }
-    } else {
-      log_warning(REQUESTER_COMPONENT_LOGGER_ID,
-                  "Transactions requests list is full\n");
+
+  rw_lock_handle_wrlock(&state->lock);
+
+  if (is_milestone) {
+    // TODO remove hash from transactions set
+    if ((ret = hash243_set_add(&state->milestones, hash)) != RC_OK) {
+      goto done;
+    }
+  } else if (!hash243_set_contains(&state->milestones, hash) &&
+             hash243_set_size(&state->transactions) < MAX_TX_REQ_NBR) {
+    if ((ret = hash243_set_add(&state->transactions, hash)) != RC_OK) {
+      goto done;
     }
   }
-  return RC_OK;
+
+done:
+  rw_lock_handle_unlock(&state->lock);
+  return ret;
 }
 
 retcode_t get_transaction_to_request(requester_state_t *const state,
-                                     trit_array_p *hash) {
+                                     flex_trit_t *const hash,
+                                     bool const milestone) {
   retcode_t ret = RC_OK;
-  concurrent_list_node_trit_array_p *iter = NULL;
+  hash243_set_t request_set = NULL;
+  hash243_set_t backup_set = NULL;
+  hash243_set_entry_t *iter = NULL;
+  hash243_set_entry_t *tmp = NULL;
+  trit_array_t key = {.trits = hash,
```

Comment:
Should assign `.trits = NULL`

Reply:
I'm actually wrapping the `flex_trit_t *` in a `trit_array_t` to do the database request

Reply:
i see, but you don't need that first assignment because of line 179,
just so it will be little more clear

Reply:
You are indeed totally right, I thought you were talking about `request_transaction`, I'll change that.